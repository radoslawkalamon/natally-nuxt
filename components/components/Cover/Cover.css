.cover {
  background-color: var(--color-betwixt);
  background-position: center;
}
