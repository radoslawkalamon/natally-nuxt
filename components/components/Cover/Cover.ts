import Vue from 'vue'

export default Vue.extend({
  name: 'ComponentsCover',
  props: {
    images: {
      type: Array as Vue.PropType<String[]>,
      default: () => []
    }
  },
  computed: {
    style (): Record<string, string> {
      if (this.images.length === 0) {
        return {}
      }

      const imageSets = this.images.map((image, index) => `url("${image}") ${index + 1}x`)
      return {
        backgroundImage: `-webkit-image-set(${imageSets.join(', ')})`
      }
    }
  }
})
