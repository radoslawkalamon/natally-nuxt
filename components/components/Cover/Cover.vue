<template>
  <div
    aria-hidden="true"
    class="cover"
    data-test="components-cover"
    :style="style"
  />
</template>

<script src="./Cover.ts" lang="ts"></script>
<style src="./Cover.css" scoped></style>
