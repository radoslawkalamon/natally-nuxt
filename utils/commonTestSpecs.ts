import Vue from 'vue'
import { mount } from '@vue/test-utils'
import { merge } from 'lodash'

export function shallRender (component: typeof Vue, dataTestID: string, defaultOptions?: object): void {
  test('shall render', (): void => {
    const wrapper = mount(component, defaultOptions)
    const isWrapperVisible = wrapper.find(`[data-test="${dataTestID}"]`).isVisible()
    expect(isWrapperVisible).toBeTruthy()
  })
}

export function shallHaveSlot (component: typeof Vue, slot: string, defaultOptions?: object): void {
  test(`shall have slot '${slot}'`, (): void => {
    const text = `This is slot '${slot}' test.`
    const wrapperOptions = merge(defaultOptions, {
      slots: {
        [slot]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHaveStringProp (component: typeof Vue, propName: string, defaultOptions?: object): void {
  test(`shall have string prop '${propName}'`, (): void => {
    const text = `This is prop ${propName} test.`
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: text
      }
    })
    const wrapper = mount(component, wrapperOptions)
    expect(wrapper.html()).toContain(text)
  })
}

export function shallHavePropInfluenceOverClassNames (
  component: typeof Vue,
  propName: string,
  propValue: unknown,
  classNameSuffix: string,
  defaultOptions?: object
) {
  test(`shall have ${typeof propValue} prop '${propName}' influence over class names`, () => {
    const wrapperOptions = merge(defaultOptions, {
      propsData: {
        [propName]: propValue
      }
    })
    const wrapper = mount(component, wrapperOptions)
    const isWrapperHaveClass = wrapper.classes().some(cls => cls.includes(classNameSuffix))
    expect(isWrapperHaveClass).toBeTruthy()
  })
}
