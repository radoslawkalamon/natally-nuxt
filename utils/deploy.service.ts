/* eslint-disable no-console */
import { Client, FileInfo, FTPResponse } from 'basic-ftp'
import { ProgressInfo } from 'basic-ftp/dist/ProgressTracker'
import capitalize from './formatters/capitalize'
import spacelize from './formatters/spacelize'

type DeployServiceConfig = {
  host: string;
  port: number;
  username: string;
  password: string;
  remoteCatalog: string;
  remoteIgnoreFile: string[];
  localCatalog: string;
}

type RemoveStrategies = {
  [key in FileInfo['type']]: (name: FileInfo['name']) => Promise<void | FTPResponse>
}

function decoratorLogCall () {
  return function (_target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    const targetMethod = descriptor.value

    descriptor.value = function (...args: any[]) {
      const formattedName = spacelize(capitalize(propertyKey))
      console.log(`::: ${formattedName} :::`)
      return targetMethod.apply(this, args)
    }

    return descriptor
  }
}

class DeployService {
  readonly host: string;
  readonly port: number;
  readonly username: string;
  readonly password: string;
  readonly remoteCatalog: string;
  readonly remoteIgnoreFile: string[];
  readonly localCatalog: string;
  FTPClient = new Client()

  constructor (config: DeployServiceConfig) {
    this.host = config.host
    this.port = Number(config.port)
    this.username = config.username
    this.password = config.password
    this.remoteCatalog = config.remoteCatalog
    this.remoteIgnoreFile = config.remoteIgnoreFile
    this.localCatalog = config.localCatalog
  }

  public async start (): Promise<void> {
    try {
      await this.openConnection()
      await this.ensureRemoteCatalog()
      await this.openRemoteCatalog()
      await this.cleanRemoteCatalog()
      await this.uploadToRemoteCatalog()
    } catch (e) {
      console.error(e)
      throw e
    } finally {
      this.closeConnection()
    }
  }

  @decoratorLogCall()
  private async openConnection (): Promise<void> {
    await this.FTPClient.access({
      host: this.host,
      port: this.port,
      user: this.username,
      password: this.password,
      secure: false
    })
  }

  @decoratorLogCall()
  private async ensureRemoteCatalog (): Promise<void> {
    await this.FTPClient.ensureDir(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async openRemoteCatalog (): Promise<void> {
    await this.FTPClient.cd(this.remoteCatalog)
  }

  @decoratorLogCall()
  private async cleanRemoteCatalog (): Promise<void> {
    const remoteFiles = await this.FTPClient.list()

    for (const file of remoteFiles) {
      if (this.remoteIgnoreFile.includes(file.name)) {
        continue
      }

      const removeStrategies: RemoveStrategies = {
        0: async () => await Promise.resolve(),
        1: async name => await this.FTPClient.remove(name),
        2: async name => await this.FTPClient.removeDir(name),
        3: async () => await Promise.resolve()
      }
      await removeStrategies[file.type](file.name)
    }
  }

  @decoratorLogCall()
  private async uploadToRemoteCatalog (): Promise<void> {
    this.FTPClient.trackProgress(this.handlerUploadProgress)
    await this.FTPClient.uploadFromDir(this.localCatalog)
    this.FTPClient.trackProgress()
  }

  private handlerUploadProgress (info: ProgressInfo): void {
    if (info.bytes === 0) {
      process.stdout.write(`Uploading file ${info.name}\n`)
    }
  }

  @decoratorLogCall()
  private closeConnection (): void {
    this.FTPClient.close()
  }
}

export default DeployService
