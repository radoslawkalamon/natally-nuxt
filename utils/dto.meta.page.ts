export type DTOMetaPageConstructor = {
  createdAt: string;
  description: string;
  imageOpenGraph?: string;
  path: string;
  title: string;
  updatedAt: string;
}

const getPath = (path: string) => {
  const pathWithoutIndex = path.split('/').filter(i => i !== 'index').join('/')
  return `${pathWithoutIndex}${pathWithoutIndex.at(-1) !== '/' ? '/' : ''}`
}

export class DTOMetaPage {
  readonly createdAt: Date;
  readonly description: string;
  readonly imageOpenGraph: string;
  readonly path: string;
  readonly title: string;
  readonly updatedAt: Date;

  constructor ({
    createdAt,
    description,
    imageOpenGraph = 'common.opengraph.png',
    title,
    updatedAt,
    path
  }: DTOMetaPageConstructor) {
    this.createdAt = new Date(createdAt)
    this.description = description
    this.imageOpenGraph = `/images/${imageOpenGraph}`
    this.path = getPath(path)
    this.title = title
    this.updatedAt = new Date(updatedAt)
  }
}
