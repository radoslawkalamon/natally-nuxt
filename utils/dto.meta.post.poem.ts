import { DTOMetaPage, DTOMetaPageConstructor } from '@/utils/dto.meta.page'

export type DTOMetaPostPoemConstructor = DTOMetaPageConstructor & {
  imageCover: string
  imageOpenGraph: string
}

export class DTOMetaPostPoem extends DTOMetaPage {
  readonly imageCover: string;
  readonly imageOpenGraph: string;

  constructor ({
    createdAt,
    description,
    imageCover,
    imageOpenGraph,
    path,
    title,
    updatedAt
  }: DTOMetaPostPoemConstructor) {
    super({
      createdAt,
      description,
      path,
      title,
      updatedAt
    })

    this.imageCover = `${this.imagePath}${imageCover}`
    this.imageOpenGraph = `${this.imagePath}${imageOpenGraph}`
  }

  get imagePath (): string {
    return '/images/poezja-314/'
  }
}
