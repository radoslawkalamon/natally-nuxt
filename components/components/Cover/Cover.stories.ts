import { Meta, Story } from '@storybook/vue'
import Cover from './Cover.vue'

const meta: Meta = {
  title: 'Components / Cover',
  component: Cover
}
export default meta

const Template: Story = (_args, { argTypes }) => ({
  components: { Cover },
  props: Object.keys(argTypes),
  template: '<Cover v-bind="$props" style="height: 300px; width: 300px;" />'
})

export const OneRatio: Story = Template.bind({})
OneRatio.args = {
  images: [
    'https://www.fillmurray.com/1200/600'
  ]
}

export const TwoRatio: Story = Template.bind({})
TwoRatio.args = {
  images: [
    'https://www.fillmurray.com/600/350',
    'https://www.fillmurray.com/1200/600'
  ]
}
