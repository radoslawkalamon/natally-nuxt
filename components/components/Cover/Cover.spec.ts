import Cover from '@/components/components/Cover/Cover.vue'
import { shallPassUnitSanityTest } from '@/devtools/jest.common.spec'
import { createUnitTestWrapper } from '@/devtools/jest.common.spec.utils'

describe('Components / Cover', () => {
  describe('Unit', () => {
    shallPassUnitSanityTest({
      component: Cover
    })

    test('shall vm.style output empty object with 0 image', async () => {
      const wrapper = await createUnitTestWrapper({
        component: Cover
      })
      // @ts-ignore
      expect(wrapper.vm.style).toMatchSnapshot()
      wrapper.destroy()
    })

    test('shall vm.style output image-set with 1 image', async () => {
      const wrapper = await createUnitTestWrapper({
        component: Cover,
        options: {
          propsData: {
            images: ['/images/post.test-post.1x.jpg']
          }
        }
      })
      // @ts-ignore
      expect(wrapper.vm.style).toMatchSnapshot()
      wrapper.destroy()
    })

    test('shall vm.style output image-set with 2 image', async () => {
      const wrapper = await createUnitTestWrapper({
        component: Cover,
        options: {
          propsData: {
            images: [
              '/images/post.test-post.1x.jpg',
              '/images/post.test-post.2x.jpg'
            ]
          }
        }
      })
      // @ts-ignore
      expect(wrapper.vm.style).toMatchSnapshot()
      wrapper.destroy()
    })
  })
})
