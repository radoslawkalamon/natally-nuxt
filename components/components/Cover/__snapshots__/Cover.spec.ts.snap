// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / Cover Unit shall pass unit sanity test 1`] = `"<div aria-hidden="true" class="cover"></div>"`;

exports[`Components / Cover Unit shall vm.style output empty object with 0 image 1`] = `{}`;

exports[`Components / Cover Unit shall vm.style output image-set with 1 image 1`] = `
{
  "backgroundImage": "-webkit-image-set(url("/images/post.test-post.1x.jpg") 1x)",
}
`;

exports[`Components / Cover Unit shall vm.style output image-set with 2 image 1`] = `
{
  "backgroundImage": "-webkit-image-set(url("/images/post.test-post.1x.jpg") 1x, url("/images/post.test-post.2x.jpg") 2x)",
}
`;
